kimchi/src/prover.rs

```diff
@@ -424,79 +395,175 @@ where
+                    .interpolate()
+                    .divide_by_vanishing_poly(index.cs.domain.d1)
+                    .unwrap();
+                assert!(res.is_zero());
+            }
+
+            drop(mul8);
+
+            // endoscaling
+            let emul8 = endosclmul::constraint(2 + range::ENDML.start).evaluations(&env);
+            t8 += &emul8;
+
+            if cfg!(test) {
+                let (_, res) = emul8
+                    .clone()
+                    .interpolate()
+                    .divide_by_vanishing_poly(index.cs.domain.d1)
+                    .unwrap();
+                assert!(res.is_zero());
+            }
+
+            drop(emul8);
+
+            // endoscaling scalar computation
+            let emulscalar8 =
+                endomul_scalar::constraint(range::ENDOMUL_SCALAR.start).evaluations(&env);
+            t8 += &emulscalar8;
+
+            if cfg!(test) {
+                let (_, res) = emulscalar8
+                    .clone()
+                    .interpolate()
+                    .divide_by_vanishing_poly(index.cs.domain.d1)
+                    .unwrap();
+                assert!(res.is_zero());
+            }
+
+            drop(emulscalar8);
+
+            // poseidon
+            let pos8 = poseidon::constraint().evaluations(&env);
+            t8 += &pos8;
+
+            if cfg!(test) {
+                let (_, res) = pos8
+                    .clone()
+                    .interpolate()
+                    .divide_by_vanishing_poly(index.cs.domain.d1)
+                    .unwrap();
+                assert!(res.is_zero());
+            }
+
+            drop(pos8);
+
+            // chacha
+            if index.cs.chacha8.as_ref().is_some() {
                 let chacha = chacha::constraint(range::CHACHA.start).evaluations(&env);
                 t4 += &chacha;
-                drop(chacha);
-                t4
+
+                if cfg!(test) {
+                    let (_, res) = chacha
+                        .clone()
+                        .interpolate()
+                        .divide_by_vanishing_poly(index.cs.domain.d1)
+                        .unwrap();
+                    assert!(res.is_zero());
+                }
             }
-        };
 
-        // quotient polynomial for lookup
-        let (t4, t8) = match index.cs.lookup_constraint_system.as_ref() {
-            None => (t4, t8),
-            Some(lcs) => combine_evaluations(
-                (t4, t8),
-                alpha,
-                alphas[alphas.len() - 1],
-                lookup::constraints(&lcs.dummy_lookup_values[0], index.cs.domain.d1)
-                    .iter()
-                    .map(|e| e.evaluations(&env))
-                    .collect(),
-            ),
-        };
+            // lookup
+            if let Some(lcs) = index.cs.lookup_constraint_system.as_ref() {
+                let mut alpha_pow = alphas[alphas.len() - 1];
+                let pows = (0..).map(|_| {
```

Comment:
It makes more sense (to me, at least) to do this in the loop below rather than constructing an iterator.

Reply:
are you commenting on the power of alpha thing or the stuff below? For the powers of alpha it doesn't matter too much as I'm going to remove this in the alpha iterator PR

Reply:
note that I'm going to remove this code in the alpha iterator PR

Reply:
also this code is not correct, the `(0..)` is going to time out. I guess we don't have tests for lookup?

Reply:
which prompted me to check wth was happening, and https://github.com/o1-labs/proof-systems/pull/289

thanks for catching these :D